### PR TITLE
Ensure type ascriptions are correctly transformed to facilitate checking of more complex return types

### LINKED ISSUE #9980: Return type is not checked if it contains an error context (`!`)
I just found out we forgot to handle `!` in our type ascriptions. For arguments that wasn't really used (even if allowed), so it was OK, but once we started using the same logic in return type checks that started to cause problems - if you add `!` to your return type ascription (which is a good practice from perspective of documentation), the type check gets disabled.

Repro:
```
from Standard.Base import all

type My_Error

f1 x -> Integer ! My_Error = x
f2 x -> Integer = x

main =
    IO.println (f1 42)
    IO.println (f2 42)
    IO.println (Panic.recover Any (f1 "txt"))
    IO.println (Panic.recover Any (f2 "txt"))
```

### Actual behaviour

```
42
42
txt
(Error: Type error: expected the result of `f2` to be Integer, but got Text.)
```

### Expected behaviour

```
42
42
(Error: Type error: expected the result of `f1` to be Integer, but got Text.)
(Error: Type error: expected the result of `f2` to be Integer, but got Text.)
```

### PR DESCRIPTION
**Pull request by radeusgd**
### Pull Request Description

- Fixes #9980
- Adds some tests to ensure types like `|` or `&` (in addition to `!` from the ticket) correctly work in return type check.
- Fixes a weird behaviour where we used to avoid processing type related IR transformations inside of type ascriptions.
- Adds parentheses to type representations if they are more complex: `A | B & C` is unclear as it can either mean `A | (B & C)` or `(A | B) & C` which have different meanings. If we now have an operation with such nesting, the sub expressions are wrapped in parentheses to disambiguate.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.

### CODE REVIEW

File: engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/resolve/TypeFunctions.scala [hunk 1]
```diff
@@ -106,16 +106,14 @@ case object TypeFunctions extends IRPass {
     * @return `expr`, with any typing functions resolved
     */
   def resolveExpression(expr: Expression): Expression = {
-    expr.transformExpressions {
-      case asc: Type.Ascription => asc
```
Comment (radeusgd) [inline engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/resolve/TypeFunctions.scala:110]: This line was causing `transformExpressions` to _not_ go deeper into the `Type.Ascription`.

I'm not sure why it was there. It was breaking the processing of transforming the expression `a | b` or `a ! b` from `Operator.Binary` into `Application.Prefix` and later into `type.Set.Union` or `Type.Error` which meant later passes were not detecting these constructs correctly.

Comment (radeusgd) [reply engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/resolve/TypeFunctions.scala:110]: I hope changing that does not break any other tests 😅

**Review by JaroslavTulach — APPROVED** [review]
Fix in `extractAscribedType` seems correct.

File: engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/resolve/TypeFunctions.scala (hunk 1 above)
Comment (radeusgd) [reply engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/resolve/TypeFunctions.scala:110]: Actually it **did break** one test.

My changes to make this test pass again are in commit aeff369af7818ca673b75f1f62d916b3cc630148

@JaroslavTulach would you mind to take a look? As I think you wrote this test originally.
What changed is that the `arg2` is no longer an `Application.Prefix` but now the `Literal`, _but_ it does have the correct type ascription - so from my perspective this behaviour is still valid, just slightly different. But please double check if I did not misinterpret the true intention behind this test.

**Review by jdunkerley — APPROVED** [review]

**Pull request merged**